File: LeetCode/Liang/24SwapNodesInPairs.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* swapPairs(ListNode* head) {
        if (!head || !head->next) {
            return head;
        }
        
        ListNode *preHead = new ListNode(INT_MIN);
        preHead->next = head;
        ListNode *pre = preHead;
        ListNode *second = pre->next->next;
        while (second) {
            //swap
            pre->next->next = second->next;
            second->next = pre->next;
            pre->next = second;
            
            //update
            pre = pre->next->next;
            second = pre->next;
            if (!second) {
                // The last pair only have one
                break;
            }
            second = second->next;
        }
        
        head = preHead->next;
        delete preHead;
        return head;
    }
};

File: LeetCode/Kai/24SwapNodesInPairs.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode swapPairs(ListNode head) {
        if(head==null || head.next==null) return head; //corner case
        
        ListNode dummy = new ListNode(0);
        dummy.next = head;
        ListNode curr = dummy;
        while(curr.next!=null && curr.next.next!=null) {
            ListNode p1 = curr.next;
            ListNode p2 = curr.next.next;
            p1.next = p2.next;
            p2.next = p1;
            curr.next = p2;
            curr = curr.next.next;
        }
        return dummy.next;
    }
    
}